pkg/collector/corechecks/containers/generic/processor_network.go

```diff
@@ -0,0 +1,117 @@
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"
+	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
+)
+
+type containerNetwork struct {
+	stats       *metrics.ContainerNetworkStats
+	containerID string
+	tags        []string
+}
+
+// ProcessorNetwork is a Processor extension taking care of network metrics
+type ProcessorNetwork struct {
+	sender                  SenderFunc
+	aggSender               aggregator.Sender
+	groupedContainerNetwork map[uint64][]containerNetwork
+}
+
+// NewProcessorNetwork returns a default ProcessorExtension
+func NewProcessorNetwork() ProcessorExtension {
+	return &ProcessorNetwork{}
+}
+
+// PreProcess is called once during check run, before any call to `Process`
+func (pn *ProcessorNetwork) PreProcess(sender SenderFunc, aggSender aggregator.Sender) {
+	pn.sender = sender
+	pn.aggSender = aggSender
+	pn.groupedContainerNetwork = make(map[uint64][]containerNetwork)
+}
+
+// Process stores each container in relevant network group
+func (pn *ProcessorNetwork) Process(tags []string, container *workloadmeta.Container, collector provider.Collector, cacheValidity time.Duration) {
+	containerNetworkStats, err := collector.GetContainerNetworkStats(container.ID, cacheValidity)
+	if err != nil {
+		log.Debugf("Gathering network metrics for container: %v failed, metrics may be missing, err: %w", container, err)
+		return
+	}
+
+	if containerNetworkStats != nil {
+		// Do not report metrics for containers running in Host network mode
+		if containerNetworkStats.UsingHostNetwork != nil && *containerNetworkStats.UsingHostNetwork {
+			return
+		}
+
+		var groupID uint64
+		if containerNetworkStats.NetworkIsolationGroupID != nil {
+			groupID = *containerNetworkStats.NetworkIsolationGroupID
+		} else {
+			groupID = 0 // Not grouped
+		}
+
+		pn.groupedContainerNetwork[groupID] = append(pn.groupedContainerNetwork[groupID], containerNetwork{stats: containerNetworkStats, containerID: container.ID, tags: tags})
+	}
+}
+
+// PostProcess actually computes the metrics
+func (pn *ProcessorNetwork) PostProcess() {
+	pn.processGroupedContainerNetwork()
+}
+
+func (pn *ProcessorNetwork) processGroupedContainerNetwork() {
+	// Handle all containers that we could not linked to any network isolation group
+	if noGroupContainers, found := pn.groupedContainerNetwork[0]; found {
+		for _, containerNetwork := range noGroupContainers {
+			pn.generateNetworkMetrics(containerNetwork.tags, containerNetwork.stats)
+		}
+	}
+	delete(pn.groupedContainerNetwork, 0)
+
+	for _, containerNetworks := range pn.groupedContainerNetwork {
+		if len(containerNetworks) == 1 {
+			pn.generateNetworkMetrics(containerNetworks[0].tags, containerNetworks[0].stats)
+		} else {
+			// If we have multiple containers, we cannot tag with HighCardinality, so re-tagging with Orchestrator card.
```

Comment:
Mind elaborating more here on why we cannot tag each container separately?

Reply:
Because the stats we gather are by network namespace. The whole story about this code is to generate "accurate" network metric.

Adding container tags mean that a query with `container_id:foo` will return a result. You may thus expect that the network traffic is due to `container_id:foo`, but, in fact, we don't know. The only thing we know is that the POD produced that much network traffic.

Reply:
Yes makes sense, thanks for the explanation! would be nice to have this brief explanation in a comment as well 🙇